src/main/java/org/springframework/data/repository/query/FunctionsMap.java

```diff
@@ -0,0 +1,122 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.springframework.core.convert.TypeDescriptor;
+import org.springframework.data.repository.query.spi.Function;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.MultiValueMap;
+
+import lombok.Data;
+
+/**
+ * {@link MultiValueMap} like data structure to keep lists of
+ * {@link org.springframework.data.repository.query.spi.Function}s indexed by name and argument list length, where the
+ * value lists are actually unique with respect to the signature.
+ *
+ * @author Jens Schauder
+ * @since 2.0
+ */
+class FunctionsMap {
+
+	private final MultiValueMap<NameAndArgumentCount, Function> functions = CollectionUtils.toMultiValueMap(new HashMap<>());
+
+	void addAll(Map<String, Function> newFunctions) {
+
+		newFunctions.forEach((n, f) -> {
+			NameAndArgumentCount k = new NameAndArgumentCount(n, f.getParameterCount());
+			List<Function> currentElements = get(k);
+			if (!contains(currentElements, f)) {
+				functions.add(k, f);
+			}
+		});
+	}
+
+	void addAll(MultiValueMap<NameAndArgumentCount, Function> newFunctions) {
+
+		newFunctions.forEach((k, list) -> {
+			List<Function> currentElements = get(k);
+			list.forEach(f -> {
+				if (!contains(currentElements, f)) {
+					functions.add(k, f);
+				}
+			});
+		});
+	}
+
+	List<Function> get(NameAndArgumentCount key) {
+		return functions.getOrDefault(key, Collections.emptyList());
+	}
+
+	/**
+	 * Gets the function that best matches the parameters given.
+	 *
+	 * The {@code name} must match, and the {@code argumentTypes} must be compatible with parameter list of the function.
+	 * In order to resolve ambiguity it checks for a method with exactly matching parameter list.
+	 *
+	 * @param name the name of the method
+	 * @param argumentTypes types of arguments that the method must be able to accept
+	 * @return a {@code Function} if a unique on gets found. {@code Optional.empty} if none matches.
+	 *    Throws {@link IllegalStateException} if multiple functions match the parameters.
+	 */
+	Optional<Function> get(String name, List<TypeDescriptor> argumentTypes) {
+
+		Stream<Function> candidates = get(new NameAndArgumentCount(name, argumentTypes.size()))
+				.stream() //
+				.filter(f -> f.supports(argumentTypes));
+		return bestMatch(candidates.collect(Collectors.toList()), argumentTypes);
+	}
+
+	private static boolean contains(List<Function> elements, Function f) {
+		return elements.stream().anyMatch(f::isSignatureEqual);
+	}
+
+	private static Optional<Function> bestMatch(List<Function> candidates, List<TypeDescriptor> argumentTypes) {
+
+		if (candidates.isEmpty()) {
+			return Optional.empty();
+		}
+		if (candidates.size() == 1) {
+			return Optional.of(candidates.get(0));
+		}
+
+		Optional<Function> exactMatch = candidates.stream().filter(f -> f.supportsExact(argumentTypes)).findFirst();
+		if (exactMatch.isPresent()) {
```

Comment:
Correct: multiple exact matches are not possible at this point because we drop those when we add elements to `Functions` this ensures the behavior of overwriting of methods from different sources (extension, root object, and aliases).

But we only check for an exact match if we have more than one match (candidates). So we throw the exception if we have multiple candidates and can't disambiguate between them.

We could improve the message to: 
> There are multiple matching methods for arguments of type <argumentTypeList>, but no exact match. Make sure to provide only one matching overload or one with exactly those types.

What do you think?